src/Messages.jl

```diff
@@ -0,0 +1,555 @@
+        end
+        return r
+    end
+end
+
+Response(bytes) = parse(Response, bytes)
+
+function reset!(r::Response)
+    r.version = v"1.1"
+    r.status = 0
+    if !isempty(r.headers)
+        empty!(r.headers)
+    end
+    if !isempty(r.body)
+        empty!(r.body)
+    end
+end
+
+
+"""
+    Request <: Message
+
+Represents a HTTP Request Message.
+
+- `method::String`
+- `uri::String`
+- `version::VersionNumber`
+- `headers::Vector{Pair{String,String}}`
+- `body::Vector{UInt8}`
+- `response`, the `Response` to this `Request`
+- `parent`, the `Response` (if any) that led to this request
+  (e.g. in the case of a redirect).
+"""
+
+mutable struct Request <: Message
+    method::String
+    uri::String
+    version::VersionNumber
+    headers::Headers
+    body::Vector{UInt8}
+    response::Response
+    parent
+end
+
+Request() = Request("", "")
+
+function Request(method::String, uri, headers=[], body=UInt8[]; parent=nothing)
+    r = Request(method,
+                uri == "" ? "/" : uri,
+                v"1.1",
+                mkheaders(headers),
+                body,
+                Response(),
+                parent)
+    r.response.request = r
+    return r
+end
+
+Request(bytes) = parse(Request, bytes)
+
+mkheaders(h::Headers) = h
+mkheaders(h)::Headers = Header[string(k) => string(v) for (k,v) in h]
+
+"""
+    issafe(::Request)
+
+https://tools.ietf.org/html/rfc7231#section-4.2.1
+"""
+
+issafe(r::Request) = r.method in ["GET", "HEAD", "OPTIONS", "TRACE"]
+
+
+"""
+    isidempotent(::Request)
+
+https://tools.ietf.org/html/rfc7231#section-4.2.2
+"""
+
+isidempotent(r::Request) = issafe(r) || r.method in ["PUT", "DELETE"]
+
+
+"""
+    iserror(::Response)
+
+Does this `Response` have an error status?
+"""
+
+iserror(r::Response) = r.status != 0 && r.status != 100 && r.status != 101 &&
+                       (r.status < 200 || r.status >= 300) && !isredirect(r)
+
+
+"""
+    isredirect(::Response)
+
+Does this `Response` have a redirect status?
+"""
+isredirect(r::Response) = r.status in (301, 302, 307, 308)
+
+
+"""
+    statustext(::Response) -> String
+
+`String` representation of a HTTP status code. e.g. `200 => "OK"`.
+"""
+
+statustext(r::Response) = Base.get(Parsers.STATUS_CODES, r.status, "Unknown Code")
+
+
+"""
+    header(::Message, key [, default=""]) -> String
+
+Get header value for `key` (case-insensitive).
+"""
+header(m::Message, k, d="") = header(m.headers, k, d)
+header(h::Headers, k::String, d::String="") = getbyfirst(h, k, k => d, lceq)[2]
+lceq(a,b) = lowercase(a) == lowercase(b)
+
+
+"""
+    hasheader(::Message, key) -> Bool
+
+Does header value for `key` exist (case-insensitive)?
+"""
+hasheader(m, k::String) = header(m, k) != ""
+
+
+"""
+    hasheader(::Message, key, value) -> Bool
+
+Does header for `key` match `value` (both case-insensitive)?
+"""
+hasheader(m, k::String, v::String) = lowercase(header(m, k)) == lowercase(v)
+
+
+"""
+    setheader(::Message, key => value)
```

Comment:
should probably be `setheader!` to match the common julia pattern

Reply:
Perhaps, `Base.setenv` could be seen as a counter-example.
I would argue that "set" in the name makes it unambiguous.
There are lost of functions in Base that modify their arguments but have no `!`. I assume this is because the verbness of the name of the function makes the side-effect obvious. e.g. `write`, `print`, `truncate`, `close`, `mark`, `reset`, etc etc.